src/withdraw-bjj.circom

```diff
@@ -0,0 +1,253 @@
+
+    // compute account state hash
+    ////////
+    component accountState = HashState();
+    accountState.tokenID <== tokenID;
+    accountState.nonce <== nonce;
+    accountState.sign <== sign;
+    accountState.balance <== balance;
+    accountState.ay <== ay;
+    accountState.ethAddr <== ethAddrState;
+    accountState.exitBalance <== exitBalance;
+    accountState.accumulatedHash <== accumulatedHash;
+
+    // verify account state is on state tree root
+    ////////
+	component smtVerify = SMTVerifier(nLevels + 1);
+	smtVerify.enabled <== 1;
+	smtVerify.fnc <== 0;
+	smtVerify.root <== rootState;
+	for (var i = 0; i < nLevels + 1; i++) {
+		smtVerify.siblings[i] <== siblingsState[i];
+	}
+	smtVerify.oldKey <== 0;
+	smtVerify.oldValue <== 0;
+	smtVerify.isOld0 <== 0;
+	smtVerify.key <== idx;
+	smtVerify.value <== accountState.out;
+
+    // check ethAddrCaller
+    var ETH_ADDR_ANY = (1<<160) - 1; // 0xFFFF...FFFF
+    component isAuthEthAddrAny = IsEqual();
+    isAuthEthAddrAny.in[0] <== ETH_ADDR_ANY;
+    isAuthEthAddrAny.in[1] <== ethAddrCallerAuth;
+
+    component checkToEthAddr = ForceEqualIfEnabled();
+    checkToEthAddr.in[0] <== ethAddrCaller;
+    checkToEthAddr.in[1] <== ethAddrCallerAuth;
+    checkToEthAddr.enabled <== 1 - isAuthEthAddrAny.out
+
+    // calculate signature hash
+    component hashSig = Poseidon(4);
+    hashSig.inputs[0] <== ethAddrCallerAuth;
+    hashSig.inputs[1] <== ethAddrBeneficiary;
+    hashSig.inputs[2] <== rootState;
+    hashSig.inputs[3] <== idx;
+
+    // computes babyjubjub X coordinate
+    component getAx = AySign2Ax();
+    getAx.ay <== ay;
+    getAx.sign <== sign;
+
+    // verify hash signature
+    component sigVerifier = EdDSAPoseidonVerifier();
+    sigVerifier.enabled <== 1;
+
+    sigVerifier.Ax <== getAx.ax;
+    sigVerifier.Ay <== ay;
+
+    sigVerifier.S <== s;
+    sigVerifier.R8x <== r8x;
+    sigVerifier.R8y <== r8y;
+    sigVerifier.M <== hashSig.out;
+
+    // compute hash global inputs
+    ////////
+    component hasherInputs = HashInputsWithdrawal(nLevels);
+
+    hasherInputs.rootState <== rootState;
+    hasherInputs.ethAddrCaller <== ethAddrCaller;
+    hasherInputs.ethAddrBeneficiary <== ethAddrBeneficiary;
+    hasherInputs.tokenID <== tokenID;
+    hasherInputs.exitBalance <== exitBalance;
+    hasherInputs.idx <== idx;
+
+    // set public output
+    hashGlobalInputs <== hasherInputs.hashInputsOut;
+
+
+}
+
+/**
+ * Computes the sha256 hash of all pretended public inputs
+ * @param nLevels - merkle tree depth
+ * @input rootState - {Field} - state root
+ * @input ethAddrCaller - {Uint160} - ethereum address that will call the SC
+ * @input ethAddrBeneficiary - {Uint160} - ethereum address that will recieve the withdraw
+ * @input tokenID - {Uint32} - token identifier
+ * @input exitBalance - {Uint192} - exit balance
+ * @input idx - {Uint48} - merkle tree index
+ * @output hashInputsOut - {Field} - hash inputs signals
+ */
+template HashInputsWithdrawal(nLevels){
```

Comment:
I would rename the template to `HashInputsWithdrawalBjj` just to distinguish it from regular withdraw